### PR TITLE
New select mock targets --no-spectra without bricks

### PR DESCRIPTION
**Pull request by forero**
This PR refactors `targets_truth_no_spectra` and solves bug #227.

### CODE REVIEW

**Comment by forero** [comment]
Some documentation and additional tests are still pending, but I leave this open for others to start reviewing the code.

**Comment by forero** [comment]
Results look much better now. The healpix boundaries are not there anymore. But there is still a glitch at the boundary for the QSOs

image
image
image
image

File: bin/mpi_select_mock_targets [hunk 1]
```diff
@@ -36,7 +36,6 @@ parser.add_argument('--realtargets', '-r', help='Path to real target catalog', t
 parser.add_argument('-v','--verbose', action='store_true', help='Enable verbose output.')
 parser.add_argument('--no-check-env', action='store_true', help="Don't check NERSC environment variables")
 parser.add_argument('--sort-pixels', action='store_true', help="Sort pixels by galactic latitude")
-parser.add_argument('--no-spectra', action='store_true')
```
Comment (sbailey) [inline bin/mpi_select_mock_targets]: For the record, from an email with @forero: `select_mock_targets --no-spectra` is now fast enough that we don't necessarily need an MPI wrapper for it.  We may want to re-enable this feature in the future to enable running the full survey in a few minutes, but ok for this to be disabled for now.

File: py/desitarget/mock/build.py [hunk 2]
```diff
@@ -1073,436 +1078,290 @@ def join_targets_truth(mockdir, outdir=None, force=False, comm=None):
...
             Output directory (default '.').
-        realtargets : astropy.table
-            Real target catalog table, e.g. from DR3 (deprecated!).
-        bricksize : float
-            Brick size for assigning bricknames, which should match the imaging
-            team value (default 0.25 deg).
         nproc : int
             Number of parallel processes to use (default 1).
-        nside : int
+        healpix_nside : int
```
Comment (sbailey) [inline py/desitarget/mock/build.py]: minor: everywhere else throughout our code we just use `nside` and not `healpix_nside`

File: py/desitarget/mock/build.py [hunk 3]
```diff
@@ -1073,436 +1078,290 @@ def join_targets_truth(mockdir, outdir=None, force=False, comm=None):
...
-    del brick_info # memory clean-up
+    return [targets, truth]
 
-    # Consolidate across all the mocks.  Note that the code quits if alltargets
-    #is zero-length, even if skytargets is non-zero length. In other words, if
-    #the parameter file only contains SKY the code will (wrongly) quit anyway.
-    if len(alltargets) == 0:
-        log.info('No targets; all done.')
-        return
+def target_selection(Selection, target_name, targets, truth, healpix_nside, healpix_id, seed, rand, log,output_dir):
```
Comment (sbailey) [inline py/desitarget/mock/build.py]: please add docstring (here and elsewhere for newly added functions)

File: py/desitarget/mock/build.py [hunk 4]
```diff
@@ -1073,436 +1078,290 @@ def join_targets_truth(mockdir, outdir=None, force=False, comm=None):
...
+        healpix_nside : int
             Healpix resolution corresponding to healpixels (default 16).
         healpixels : numpy.ndarray or int
             Restrict the sample of mock targets analyzed to those lying inside
-            this set (array) of healpix pixels.
-
+            this set (array) of healpix pixels. Default (None)
+        
     Returns:
-        A variety of fits files are written to output_dir.
```
Comment (sbailey) [inline py/desitarget/mock/build.py]: What does initialize do?  "Initializes variables..." what variables?  Does it write any files anymore?  Create directories?  Change some global variables?

**Comment by forero** [comment]
@sbailey that error message was a real bug. If you run the same command it works. It will take ~1 hour to run. If you run it with `--nside 4` it takes 5 minutes. Running time is proportional to the amount of pixels it has to process.

File: py/desitarget/mock/build.py (hunk 2 above)
Comment (forero) [reply py/desitarget/mock/build.py]: Changed.

**Comment by forero** [comment]
Latest full sky results look good
image
image
image
image
image

File: py/desitarget/mock/build.py (hunk 4 above)
Comment (forero) [reply py/desitarget/mock/build.py]: I have updated all the docstrings in the new functions.

**Comment by forero** [comment]
All comments from @sbailey are addressed/solved. This looks good to me.

**Comment by moustakas** [comment]
Will try to review this by today---thanks.

**Comment by forero** [comment]
@moustakas thanks! 
Integration tests are failing with
>/home/travis/build/desihub/desitarget/py/desitarget/mock/build.py:docstring of desitarget.mock.build.downsample_pixel:3: ERROR: Unexpected indentation.

I don't see the problem in the docstring. Do you understand why is this happening?

**Comment by sbailey** [comment]
Alas, sphinx/rst are very picky.  I think you need a blank line before "Args" and "Options" and a colon instead of a period.  e.g.
```
function description.

Args:
    blat: array
       stuff about blat...
```
instead of
```
function description
Args.
    blat: array
        stuff about blat....
```

This formatting change applies to many of your new docstrings.

To test this locally without going via travis, you can use `python setup.py build_sphinx`.  You'll need sphinx 1.5 (not 1.6) for that.

**Comment by forero** [comment]
Sphinx build is working. 
@moustakas @sbailey any new suggestions?

**Comment by forero** [comment]
> Most of the new code / refactoring in mock.build is used for the new targets_truth_nospectra function even though targets_truth needs / uses a lot of the same machinery. I guess this can all stay separate although it makes it hard for others to figure out what's going on and makes for a lot of duplicated code.

I made the refactor to allow the possibility for the next iteration in `targets_truth` to explicitly reuse those functions and move to a merge where the only thing we need in `targets_truth` is an `spectra=False` option.

> You loop on healpixels (outer loop) and then target types (inner loop) but since not all the mock catalogs are organized by healpixels it means that there's more I/O overhead than there could be. This is probably fine though since we still only read the ra,dec boundaries we need, so it's not as bad as it used to be.

It runs just fine, much better than it used to.

> Why don't you use the density fluctuations results (#217)?

Because it needs significantly new code. It will come in a new PR.

**Comment by forero** [comment]
Merging now.

**Comment by forero** [comment]
@moustakas BTW, the option for contaminants was also removed from `--no-spectra` because it uses bricks and not pixels. This will have to be reintroduced again.

**Pull request merged**